Add load weight notes (#239)
Oneflow-Inc/libai: docs/source/notes/How_to_implement_huggingface's_weights_in_LiBai.md

# How to use Huggingface's pretrained weights in LiBai

> LiBai adopts the model structures which are more suitable for parallel training, so they are a little bit different from Huggingface's ones. In this tutorial, we introduce how to correctly load Huggingface's pretrained weights with LiBai model, let's take BERT as an example.

Reviewer: ```suggestion
The built-in layers in [LiBai](https://github.com/Oneflow-Inc/libai) adopts the structure which is more suitable for parallel training, therefore there may be little bit different from the Huggingface's implementation. In this tutorial, we will introduce users how to correctly load Huggingface's pretrained weights into LiBai's models, let's take BERT as an example.
```

Reviewer: 这样比较合适

Author: 已修改

## LiBai Transformer vs Huggingface Transformer

> You can see the subtle differences in the BERT structure from the following figure (left: LiBai, right: Huggingface):

Reviewer: ```suggestion
There are subtle differences in the BERT structure in the following figure (left: LiBai, right: Huggingface), which can be summarized as:
```
感觉这样合适一些

Author: 可以的

- Location of layernorm: The location of layernorm is different, but the calculation order is the same.
- Division mode of query, key and value: Just splice.

> - LiBai follows [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) to use the order of the layernorm and the residual connections by default. Megatron-LM shows that this structure will eliminate instabilities and bring a lower training loss. LiBai can also support the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf), just set `apply_residual_post_layernorm=True`.

Reviewer: ```suggestion
- LiBai's transformer layer adopts the structure of  [Megatron-LM](https://github.com/NVIDIA/Megatron-LM)'s transformer layer. Megatron-LM shows that this structure will eliminate instabilities and bring a lower training loss. LiBai can also support the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf), just set `apply_residual_post_layernorm=True`.
```

Reviewer: 感觉这样更合适

Author: 这里要不就按原来这个吧，感觉详细一点，然后megatron的论文里也是主要解释layernorm和残差连接的不同。
![image](https://user-images.githubusercontent.com/53039617/162176292-3f6d6685-af52-4eff-99d9-42c7065a9626.png)

> - LiBai follows [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) to use the order of the layernorm and the residual connections by default. Megatron-LM shows that this structure will eliminate instabilities and bring a lower training loss. LiBai can also support the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf), just set `apply_residual_post_layernorm=True`.

Reviewer: ```suggestion
- LiBai follows [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) to use the order of the layernorm and the residual connections by default. Megatron-LM shows that this structure will eliminate instabilities and bring a lower training loss. LiBai can also support the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf) by setting `apply_residual_post_layernorm=True`.
```

![architecture](./assets/architecture.png)

## Difference of QKV slice logic
LiBai's QKV calculation logic is different from huggingface.
```python

> # LiBai's QKV calculation logic

Reviewer: 这里改为`QKV slicing logic`会不会更合适一些

Author: 也是可以的

query_key_value = query_key_value.view(batch_size, -1, num_heads, 3 * head_size)
query_key_value = query_key_value.permute(0, 2, 1, 3)
query, key, value = flow.chunk(query_key_value, chunks=3, dim=-1)

# Huggingface's QKV calculation logic
query, key, value = flow.chunk(query_key_value, chunks=3, dim=-1)
query = query.view(query.size(0), query.size(1), num_heads, -1).permute(0, 2, 1, 3)
key = key.view(key.size(0), key.size(1), num_heads, -1).permute(0, 2, 1, 3)
value = value.view(value.size(0), value.size(1), num_heads, -1).permute(0, 2, 1, 3)
```

> ## QKV weight loading method

Reviewer: ```suggestion
## How to correctly load QKV weights
```

Reviewer: 感觉这样更合适一些

Author: 好的

> - Just rearrange the weights to load correctly. 

Reviewer: ```suggestion
- To correctly load huggingface's transformer weights, the only thing you need to do is to rearrange the loaded weights as follows:
```
这样感觉更好一些

Author: 好的

```python
def convert_qkv_weight(cfg, qkv_weight, qkv_bias):
    qkv_weight = qkv_weight.view([3, cfg.num_heads, cfg.head_size, cfg.hidden_size])
    qkv_weight = qkv_weight.permute(1, 0, 2, 3).contiguous().view(3*cfg.hidden_size, cfg.hidden_size)
    qkv_bias = qkv_bias.view(3, cfg.num_heads, cfg.head_size)
    qkv_bias = qkv_bias.permute(1,0,2).contiguous().view(-1)
    return qkv_weight, qkv_bias
```

> 

Reviewer: 这里要不要再补充一段测试代码，举例一下这2个函数具体是怎么用的

Author: 感觉之前我们写的那个测试代码也不太好理解，要不直接让看这个测试吧：https://github.com/Oneflow-Inc/libai/tree/test_bert_load_huggingface_weight/projects/test_bert_load_huggingface_weight

Reviewer: 主要是感觉在某个分支下，用户会感觉怪怪的，如果代码不多的话，我感觉可以放在你的某个project下，然后合并到主分支里

Author: 也是阔以的，那要不等等我的simcse？性能已经没问题了。

Reviewer: 可以先合并，然后等在SimCSE那个分支下把这里给改了也行

> - For detailed examples, please refer to [load-huggingface-bert](https://github.com/Oneflow-Inc/libai/tree/test_bert_load_huggingface_weight/projects/test_bert_load_huggingface_weight), you can verify this by running:

Reviewer: 这是在某个具体的分支下，感觉如果不熟悉git的用户会比较尴尬

# How to use Huggingface's pretrained weights in LiBai
The built-in layers in [LiBai](https://github.com/Oneflow-Inc/libai) adopts the structure which is more suitable for parallel training, therefore there may be little bit different from the Huggingface's implementation. In this tutorial, we will introduce users how to correctly load Huggingface's pretrained weights into LiBai's model, let's take BERT as an example.

## LiBai Transformer vs Huggingface Transformer
There are subtle differences in the BERT structure in the following figure (left: LiBai, right: Huggingface), which can be summarized as:
- Location of layernorm: The location of layernorm is different, but the calculation order is the same.

> - Division mode of query, key and value: Just splice.

Reviewer: ```suggestion
- The different slicing way to get the `query`, `key` and `value` vector.
```
这样感觉好些

> # How to impllement huggingface's weights in LiBai

Reviewer 2: ```suggestion
# How to use Huggingface's pretrained weights in LiBai
```

Author: 已修改

> Because Libai's model structure has undergone rigorous parallel mode derivation, Libai'model structure is more suitable for parallel training. In this tutorial, we introduces how to correctly load huggingface's weights with Libai model, let's take BERT as an example.

Reviewer 2: ```suggestion
LiBai adopts the model structures which are more suitable for parallel training, so they are a little bit different from Huggingface's ones. In this tutorial, we introduce how to correctly load Huggingface's pretrained weights with LiBai model, let's take BERT as an example.
```

Author: 好的 已修改

> Because Libai's model structure has undergone rigorous parallel mode derivation, Libai'model structure is more suitable for parallel training. Next, we introduces how to correctly load huggingface's weights with Libai model from several aspects, and takes Bert as an example.

Reviewer: 这里感觉表达的比较口语, 应该是想表达, LiBai下的模块已经经过严谨的并行逻辑推导, 可以保证在各种并行下最优的通信和计算效率, 第一句应该改成这样对吧, 但是不太好翻译2333, 第一句可能需要其他人来看看，我也不太知道怎么表达最合适, 第二句可以这样改:
- 第二句
> Next, we introduces how to correctly load huggingface's weights with Libai model from several aspects, and takes Bert as an example.
- 改为
> In this tutorial, we instroduces how to correctly load huggingface's weights into models built with `libai.layers`, let's take BERT as an example.

Author: 好的，第二句其实是表达从各个角度来解释，我改成：In this tutorial, we introduces how to correctly load huggingface's weights with Libai model, let's take BERT as an example.

> ## Model architecture difference

Reviewer: ```suggestion
## Huggingface Transformer vs LiBai Transformer
```

Author: ok

> ## Model architecture difference

Reviewer: ```suggestion
## Huggingface Transformer vs LiBai Transformer
```

> You can see the subtle differences in the Bert structure from the following figure (left: LiBai, right: huggingface):

Reviewer: Bert都需要改为`BERT`

Author: ok

> You can see the subtle differences in the Bert structure from the following figure (left: LiBai, right: huggingface):

Reviewer: Bert都需要改为`BERT`

> You can see the subtle differences in the BERT structure from the following figure (left: LiBai, right: huggingface):

Reviewer 2: ```suggestion
You can see the subtle differences in the BERT structure from the following figure (left: LiBai, right: Huggingface):
```

Author: 已修改

- Location of layernorm: The location of layernorm is different, but the calculation order is the same.

> - Shape of query, key and value: Just splice.

Reviewer: Shape of query, key, value很奇怪，shape都是一样的，是切分方式不一样

> - Shape of query, key and value: Just splice.

Reviewer: Shape of query, key, value很奇怪，shape都是一样的，是切分方式不一样

Author: 好的 这里主要是表达两种qkv的定义形状不一致，一种是单独定义q、k、v，一种是定义qkv，已经改成：Division mode of query, key and value

> - Libai uses the order of the layer normalization and the residual connections of Megatron by default, relevant experiments have proved this order eliminates instabilities and also has a lower training loss. But LiBai also supports the original BERT architecture, just set `apply_residual_post_layernorm=True`.

Reviewer: ```suggestion
- Libai uses the order of the layer normalization and the residual connections as [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) by default, recent research shows that this structure will eliminate instabilities and lead to a lower training loss. But LiBai also supports the original BERT architecture mentioned in (这里加一个代码链接), just set `apply_residual_post_layernorm=True`.
```

Author: 好的 这里用lead to不太好 我改成bring了 表达带来一个优势

> - Libai uses the order of the layer normalization and the residual connections of Megatron by default, relevant experiments have proved this order eliminates instabilities and also has a lower training loss. But LiBai also supports the original BERT architecture, just set `apply_residual_post_layernorm=True`.

Reviewer: ```suggestion
- Libai uses the order of the layer normalization and the residual connections as [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) by default, recent research shows that this structure will eliminate instabilities and lead to a lower training loss. But LiBai also supports the original BERT architecture mentioned in (这里加一个代码链接), just set `apply_residual_post_layernorm=True`.
```

![architecture](./assets/architecture.png)

> ## QKV calculation logic difference

Reviewer: 应该这里是slice logic difference吧，差不多这样的意思，感觉不是calculation的logic，calculation的logic都是一样的

Author: 好的，改了

> ## QKV calculation logic difference

Reviewer: 应该这里是slice logic difference吧，差不多这样的意思，感觉不是calculation的logic，calculation的logic都是一样的

> - Libai uses the order of the layer normalization and the residual connections of [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) by default, recent research shows that this structure will eliminate instabilities and bring a lower training loss. But LiBai also supports the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf), just set `apply_residual_post_layernorm=True`.

Reviewer 2: layernorm 最好都用同一种写法，不要前面写 layernorm，后面写 layer normalization

Reviewer 2: ```suggestion
- LiBai follows [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) to use the order of the layernorm and the residual connections by default. Megatron-LM shows that this structure will eliminate instabilities and bring a lower training loss. LiBai can also support the original BERT architecture mentioned in [Paper](https://arxiv.org/pdf/1810.04805.pdf), just set `apply_residual_post_layernorm=True`.
```

Author: 好的 已修改 统一用layernorm

> ## QKV slice logic difference

Reviewer 2: ```suggestion
## Difference of QKV slice logic
```

Author: 已修改

> LiBai's QKV calculation logic is difference from huggingface.

Reviewer 2: ```suggestion
LiBai's QKV calculation logic is different from huggingface.
```

Author: 已修改

```python
# LiBai's QKV calculation logic
query_key_value = query_key_value.view(batch_size, -1, num_heads, 3 * head_size)
query_key_value = query_key_value.permute(0, 2, 1, 3)
query, key, value = flow.chunk(query_key_value, chunks=3, dim=-1)

# Huggingface's QKV calculation logic
query, key, value = flow.chunk(query_key_value, chunks=3, dim=-1)
query = query.view(query.size(0), query.size(1), num_heads, -1).permute(0, 2, 1, 3)
key = key.view(key.size(0), key.size(1), num_heads, -1).permute(0, 2, 1, 3)
value = value.view(value.size(0), value.size(1), num_heads, -1).permute(0, 2, 1, 3)
```

## QKV weight loading method
- Just rearrange the weights to load correctly. 

> ```python

Reviewer: 我感觉这段python代码没有太大的必要,  而且你没办法保证执行的结果都为True对吧, 给个接口函数就行了, 类似这样:
https://github.com/Oneflow-Inc/libai/blob/f8091314b2d405a8386fe25be20de9065d0b9916/projects/MAE/utils/weight_convert.py#L24

Author: 已修改

> ```python

Reviewer: 我感觉这段python代码没有太大的必要,  而且你没办法保证执行的结果都为True对吧, 给个接口函数就行了, 类似这样:
https://github.com/Oneflow-Inc/libai/blob/f8091314b2d405a8386fe25be20de9065d0b9916/projects/MAE/utils/weight_convert.py#L24

Reviewer 3: 这段代码感觉跟实际运用的时候的结构不一致, 用户可能截取或者理解都需要绕一下

这里用函数的形式包一下更容易理解
里面有 oneflow_qkv_forward(), huggingface_qkv_forward(), 以及convert_qkv_weight()

最后验证的时候, 调用 上述的几个函数,  对比一下输出. 
这样更方便理解一点, 也更贴近实际运用的场合

Author: 已修改
